File: haskell/src/Solutions/Euler166.hs
module Solutions.Euler166 where
import Utils.Operators((<&&>))

{- Idea: plug in 9 values, compute rest
abcd
efgh
ijkl
mnop
-}

isDigit = (>=0) <&&> (<10)

validSquares = [() |
    c <- [0..9],
    d <- [0..9],
    e <- [0..9],
    f <- [0..9],
    g <- [0..9],
    j <- [0..9],
    l <- [0..9],
    m <- [0..9],
    let s = d + g + j + m,
    let h = s - e - f - g,
    isDigit h,
    let p = s - d - h - l,
    isDigit p,
    o <- [0..9],
    let k = s - c - g - o,
    isDigit k,
    let i = s - j - k - l,
    isDigit i,
    let n = s - m - o - p,
    isDigit n,
    let b = s - f - j - n,
    isDigit b,
    let a = s - b - c - d,
    isDigit a,
    a + f + k + p == s,
    a + e + i + m == s]

euler166 :: IO String
euler166 = return $ show $ length validSquares

File: rust/src/solutions/euler166.rs
use itertools::iproduct;
// plug in 9 values, compute rest
// abcd
// efgh
// ijkl
// mnop

pub fn euler166() -> String {
    let mut ans = 0;
    let digitize = |d: i32| if d >= 0 && d < 10 { Some(d) } else { None };
    for (c, d, e, f, g, j, l, m) in
        iproduct!(0..10, 0..10, 0..10, 0..10, 0..10, 0..10, 0..10, 0..10)
    {
        let sum = d + g + j + m;
        if let Some(h) = digitize(sum - e - f - g) {
            if let Some(p) = digitize(sum - d - h - l) {
                for o in 0..10 {
                    if let Some(k) = digitize(sum - c - g - o) {
                        if let Some(i) = digitize(sum - j - k - l) {
                            if let Some(n) = digitize(sum - m - o - p) {
                                if let Some(b) = digitize(sum - f - j - n) {
                                    if let Some(a) = digitize(sum - b - c - d) {
                                        if a + f + k + p == sum && a + e + i + m == sum {
                                            ans += 1;
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    ans.to_string()
}
